### ISSUE TITLE
Documentation of the DumpLinearLeastSquaresProblem function

### ISSUE THREAD

**Issue by tipf** [ceres-solver/ceres-solver#608]
Hi everyone,
I tried to use the [DumpLinearLeastSquaresProblem](https://github.com/ceres-solver/ceres-solver/blob/d656d34809a7958f5b0b1ae246a3f6482e67823e/internal/ceres/linear_least_squares_problems.h#L74) function by enabling it in the solver option. I used the textfile version.
The function itself works perfectly fine, but the [documentation](http://ceres-solver.org/nnls_solving.html#_CPPv2N5ceres6Solver7Options32trust_region_problem_dump_formatE) is what makes me struggle. It says:

> The Jacobian is dumped as a text file containing (i,j,s) triplets, the vectors D, x and f are dumped as text files containing a list of their values.

The function returns some variables **A**, **D**, **x** and **b** and writes them to textfiles. It is relatively easy to figure out that **A = J** (in the documentation **J** is usually used for the Jacobian) and **b = f**.
But I can't make sense of the returned value **x**. It seem to be neither the current state **x**, nor the step **Δx** that is applied by the optimizer. I tried to sum all **x** and the result is different from the overall distance between the initial and the optimized state.
[This](https://github.com/ceres-solver/ceres-solver/blob/d656d34809a7958f5b0b1ae246a3f6482e67823e/internal/ceres/linear_least_squares_problems.h#L58) suggest that it might be the a non-damped Gauss-Newton step, is this assumption true? Wouldn't it be better to return the actual trust region step? Or do I understand something completely wrong?

Nevertheless, I suggest to improve the documentation of the DumpLinearLeastSquaresProblem function and the related solver options. At least for me, it is a bit unclear right now. :-)

**Comment by sandwichmaker** [comment]
@tipf you are right that the documentation for this is not great. In part because I have been the only user of this API for internal debugging.  Would you care to send a patch to update it?

http://ceres-solver.org/contributing.html

**Comment by tipf** [comment]
I will look into it @sandwichmaker.

**Comment by tipf** [comment]
Well, I looked through the code:

1. The meaning of the dumped **x** depends on the chosen trust region strategy:
a. In the Levenberg-Marquardt case, **x**  is the damped trust region step.
b. In the Dogleg case, **x**  is the pure Gauss-Newton step (without limiting it to the trust region radius).
2. In both cases, it is the solution of the scaled problem. The actual step that is applied by the solver is calculated [here](https://github.com/ceres-solver/ceres-solver/blob/adb973e4a337c372aa81ca1a4f3bb704068c08b7/internal/ceres/trust_region_minimizer.cc#L424).

So my idea is to remove the call to the dump function from the [LM strategy](https://github.com/ceres-solver/ceres-solver/blob/189f0c2f6efd07c3976953a5c278dbb5f263183b/internal/ceres/levenberg_marquardt_strategy.cc#L129) and the [Dogleg strategy](https://github.com/ceres-solver/ceres-solver/blob/94712db5c70116f1832fbedbfbf98f26923e8b48/internal/ceres/dogleg_strategy.cc#L576) and add it [here](https://github.com/ceres-solver/ceres-solver/blob/adb973e4a337c372aa81ca1a4f3bb704068c08b7/internal/ceres/trust_region_minimizer.cc#L386) just after the call to the trust region strategy. Additionally, I would add the Jacobian scaling to the dump function, so that the user can undo the scaling if the actual trust region step is required.
The only "drawback" is, that the dump will return the trust-region step for Dogleg and not the Gauss-Newton step. Even if it seems more consistent to return the DL step for the DL strategy, I don't now if it this behavior is preferred.

What do you think @sandwichmaker,  should I prepare a path for this?

**Comment by sandwichmaker** [comment]
Sorry about the delayed reply.

The reason we do the dumping inside the trust region strategies rather than inside the trust region minimizer is because to be able to debug problems we need access to the regularizing matrix D, and thats not accessible to the trust region minimizer.  So I do not see how you will be able to solve that.

**Comment by tipf** [comment]
You are right, I forgot that D is not accessible form outside of the individual trust region strategies. I cloud imagine to split the DumpLinearLeastSquaresProblem in two separate functions, one inside the trust region strategies and a second one outside in the ComputeTrustRegionStep code. Since there are individual files for each variable anyway, this should be possible.

However, my priorities have changed a bit in the meanwhile. I'm sorry, but I do not have the time to do this currently. I may come back to this later in October, but I cannot guarantee that I will.
